File: sdks/go/providers/xai/voices.go
package xai

import (
	"context"
	"errors"
	"net/http"
	"net/url"
	"strings"
	"unicode/utf8"

	"github.com/speechswitch/client/sdks/go/generated/auth"
	out "github.com/speechswitch/client/sdks/go/generated/xai_output"
	"github.com/speechswitch/client/sdks/go/runtime"
)

type VoiceOptions struct {
	Auth auth.Auth
	// Nil uses native HTTP; overrides must honor cancellation and reject redirects/retries/ambient credentials.
	Transport runtime.HTTPTransport
	BaseURL   string
	TimeoutMs runtime.Optional[int64]
	// Zero selects 4 MiB.
	MaxResponseBytes int
}

// Voices lists built-in voices. Existing custom IDs can be sent directly to Synthesize.
func Voices(ctx context.Context, options VoiceOptions) ([]out.Voice, error) {
	value, err := discovery(ctx, "/v1/tts/voices", options)
	if err != nil {
		return nil, err
	}
	object, ok := value.(map[string]any)
	if !ok {
		return nil, errors.New("Invalid xAI voice list")
	}
	items, ok := object["voices"].([]any)
	if !ok {
		return nil, errors.New("Invalid xAI voice list")
	}
	voices := make([]out.Voice, len(items))
	for i, item := range items {
		voices[i], err = decodeVoice(item)
		if err != nil {
			return nil, err
		}
	}
	return voices, nil
}
func Voice(ctx context.Context, id string, options VoiceOptions) (out.Voice, error) {
	if id == "" || !utf8.ValidString(id) {
		return out.Voice{}, errors.New("xAI voice ID must be a nonempty UTF-8 string")
	}
	// Keep arbitrary IDs, including dot segments and reserved characters, in one path component.
	encoded := strings.ReplaceAll(url.QueryEscape(id), "+", "%20")
	encoded = strings.ReplaceAll(encoded, ".", "%2E")
	value, err := discovery(ctx, "/v1/tts/voices/"+encoded, options)
	if err != nil {
		return out.Voice{}, err
	}
	return decodeVoice(value)
}
func decodeVoice(value any) (out.Voice, error) {
	result := out.Voice{}
	object, ok := value.(map[string]any)
	if !ok {
		return result, errors.New("Invalid xAI voice")
	}
	id, idOK := object["voice_id"].(string)
	name, nameOK := object["name"].(string)
	if !idOK || !nameOK {
		return result, errors.New("Invalid xAI voice")
	}
	result.VoiceId, result.Name = id, name
	if language, present := object["language"]; present {
		var value out.VoiceLanguage
		if language == nil {
			value = out.VoiceLanguageAsNull{}
		} else {
			text, ok := language.(string)
			if !ok {
				return result, errors.New("Invalid xAI voice language")
			}
			value = out.VoiceLanguageAsString{Value: text}
		}
		result.Language = runtime.Some(value)
	}
	return result, nil
}
func discovery(ctx context.Context, path string, options VoiceOptions) (value any, err error) {
	key, err := apiKey(options.Auth)
	if err != nil {
		return nil, err
	}
	limit := options.MaxResponseBytes
	if limit == 0 {
		limit = 4 * 1024 * 1024
	}
	if limit < 0 || uint64(limit) > 9007199254740991 {
		return nil, errors.New("xAI MaxResponseBytes must be a positive safe integer")
	}
	base := options.BaseURL
	if base == "" {
		base = "https://api.x.ai"
	}
	address, err := endpoint(base, path, false)
	if err != nil {
		return nil, err
	}
	operation, cancel, err := operationContext(ctx, options.TimeoutMs)
	if err != nil {
		return nil, err
	}
	defer cancel()
	request, err := http.NewRequestWithContext(operation, "GET", address, nil)
	if err != nil {
		return nil, err
	}
	request.Header = http.Header{"Authorization": {"Bearer " + key}, "Accept": {"application/json"}}
	transport := options.Transport
	if transport == nil {
		transport = &http.Client{CheckRedirect: func(*http.Request, []*http.Request) error { return http.ErrUseLastResponse }}
	}
	response, err := runtime.OpenResponse(request, transport)
	if err != nil {
		return nil, err
	}
	defer func() {
		closeErr := response.Body.Close()
		if err == nil {
			err = closeErr
		}
	}()
	if response.StatusCode < 200 || response.StatusCode >= 300 {
		return nil, &Error{Status: runtime.Some(response.StatusCode)}
	}
	if err := contentType(response.Header, true); err != nil {
		return nil, err
	}
	return readJSON(operation, response.Body, limit)
}

File: sdks/rust/src/providers/xai/voices.rs
use super::{content_type, credential, endpoint, failure, limit};
use crate::{
    generated::{
        auth::Auth,
        xai_output::{Voice, VoiceLanguage},
    },
    http::{HttpRequest, HttpTransport, TransportError},
    json::Raw,
};

pub struct VoiceOptions<'a> {
    pub auth: Option<&'a Auth>,
    pub transport: Option<&'a dyn HttpTransport>,
    pub base_url: Option<&'a str>,
    pub max_response_bytes: usize,
}
impl Default for VoiceOptions<'_> {
    fn default() -> Self {
        Self {
            auth: None,
            transport: None,
            base_url: None,
            max_response_bytes: 4 * 1024 * 1024,
        }
    }
}
async fn fetch(path: &str, options: VoiceOptions<'_>) -> Result<String, TransportError> {
    let key = credential(options.auth)?;
    limit(options.max_response_bytes, "max_response_bytes")?;
    let url = endpoint(options.base_url.unwrap_or("https://api.x.ai"), path, false)?;
    let mut response = options
        .transport
        .ok_or_else(|| failure("xAI HTTP transport is required"))?
        .send(HttpRequest {
            method: "GET".into(),
            url,
            headers: vec![
                ("Authorization".into(), format!("Bearer {key}")),
                ("Accept".into(), "application/json".into()),
            ],
            body: Vec::new(),
        })
        .await?;
    content_type(&response, true)?;
    let mut bytes = Vec::new();
    let mut budget = 0;
    while let Some(chunk) = std::future::poll_fn(|cx| {
        if budget == 128 {
            budget = 0;
            cx.waker().wake_by_ref();
            return std::task::Poll::Pending;
        }
        budget += 1;
        response.body.as_mut().poll_next(cx)
    })
    .await
    {
        let chunk = chunk?;
        if chunk.len() > options.max_response_bytes - bytes.len() {
            return Err(failure("xAI response exceeds max_response_bytes"));
        }
        bytes.extend(chunk);
    }
    let text = String::from_utf8(bytes).map_err(|_| failure("Invalid xAI JSON"))?;
    Ok(text.strip_prefix('\u{feff}').unwrap_or(&text).to_owned())
}
fn parse(raw: Raw<'_>) -> Result<Voice, TransportError> {
    let fields = raw.object().map_err(|_| failure("Invalid xAI voice"))?;
    let name = fields
        .get("name")
        .and_then(|v| v.string().ok())
        .ok_or_else(|| failure("Invalid xAI voice"))?;
    let voice_id = fields
        .get("voice_id")
        .and_then(|v| v.string().ok())
        .ok_or_else(|| failure("Invalid xAI voice"))?;
    let language = fields
        .get("language")
        .map(|v| {
            if v.is_null() {
                Ok(VoiceLanguage::Null(Default::default()))
            } else {
                v.string()
                    .map(VoiceLanguage::String)
                    .map_err(|_| failure("Invalid xAI voice"))
            }
        })
        .transpose()?;
    Ok(Voice {
        name,
        voice_id,
        language,
    })
}
/// Discover built-in voices. Existing custom IDs need no discovery round trip.
pub async fn voices(options: VoiceOptions<'_>) -> Result<Vec<Voice>, TransportError> {
    let text = fetch("/v1/tts/voices", options).await?;
    let raw = Raw::parse_exact(&text).map_err(|_| failure("Invalid xAI JSON"))?;
    let fields = raw
        .object()
        .map_err(|_| failure("Invalid xAI voices response"))?;
    let values = fields
        .get("voices")
        .and_then(|v| v.array().ok())
        .ok_or_else(|| failure("Invalid xAI voices response"))?;
    values.into_iter().map(parse).collect()
}
pub async fn voice(id: &str, options: VoiceOptions<'_>) -> Result<Voice, TransportError> {
    if id.is_empty() {
        return Err(failure("xAI voice ID must not be empty"));
    }
    let mut encoded = String::new();
    for byte in id.bytes() {
        if byte.is_ascii_alphanumeric() || matches!(byte, b'-' | b'_' | b'~') {
            encoded.push(byte as char);
        } else {
            use std::fmt::Write;
            write!(encoded, "%{byte:02X}").expect("String write");
        }
    }
    let text = fetch(&format!("/v1/tts/voices/{encoded}"), options).await?;
    parse(Raw::parse_exact(&text).map_err(|_| failure("Invalid xAI JSON"))?)
}
